port/unix/omrsysinfo.c

```diff
@@ -6439,6 +6505,75 @@ getCgroupMemoryLimit(struct OMRPortLibrary *portLibrary, uint64_t *limit)
 	return rc;
 }
 
+/**
+ * Get the cgroup subsystem metric map and its number of elements given the
+ * subsystem flag and the cgroup version in use.
+ *
+ * @param[in] portLibrary pointer to OMRPortLibrary
+ * @param[in] subsystem cgroup subsystem flag
+ * @param[out] subsystemMetricMap if not null, on successful return is assigned
+ *      the metric map associated with the subsystem flag
+ * @param[out] numElements if not null, on successful return is assigned the
+ *      number of elements of the metric map associated with the subsystem flag
+ *
+ * @return 0 on success, otherwise negative error code
+ */
+static int32_t
+getCgroupSubsystemMetricMap(struct OMRPortLibrary *portLibrary, uint64_t subsystem, const struct OMRCgroupSubsystemMetricMap **subsystemMetricMap, uint32_t *numElements)
+{
+	const struct OMRCgroupSubsystemMetricMap *map = NULL;
+	uint32_t count = 0;
+	int32_t rc = 0;
+
+	Assert_PRT_true((NULL != subsystemMetricMap) || (NULL != numElements));
+	if (OMR_ARE_ANY_BITS_SET(PPG_sysinfoControlFlags, OMRPORT_SYSINFO_CGROUP_V1_AVAILABLE)) {
+		switch (subsystem) {
+		case OMR_CGROUP_SUBSYSTEM_MEMORY:
+			map = omrCgroupMemoryMetricMapV1;
+			count = OMR_CGROUP_MEMORY_METRIC_MAP_V1_SIZE;
+			break;
+		case OMR_CGROUP_SUBSYSTEM_CPU:
+			map = omrCgroupCpuMetricMapV1;
+			count = OMR_CGROUP_CPU_METRIC_MAP_V1_SIZE;
+			break;
+		case OMR_CGROUP_SUBSYSTEM_CPUSET:
+			map = omrCgroupCpusetMetricMapV1;
+			count = OMR_CGROUP_CPUSET_METRIC_MAP_V1_SIZE;
+			break;
+		default:
+			rc = OMRPORT_ERROR_SYSINFO_CGROUP_SUBSYSTEM_UNAVAILABLE;
+		}
+	} else if (OMR_ARE_ANY_BITS_SET(PPG_sysinfoControlFlags, OMRPORT_SYSINFO_CGROUP_V2_AVAILABLE)) {
```

Comment:
Reading online, it seems that both v1 and v2 controllers can be mounted on the same system. As the change stands, it looks like OMR will favour v1 (based on what I would expect the `ARE_ANY_BITS` query to return); is that the desired behaviour?

Reply:
Yes; we check for v1 by seeing if the mount point (/sys/fs/cgroup) is tmpfs and we check for v2 by looking for `cgroup.controllers` file at the mount point. (/sys/fs/cgroup is the default mount point, see https://github.com/eclipse/omr/issues/1281#issuecomment-1076400298)
So either the mount point is `tmpfs` or it is `cgroup2fs` and contains the cgroup.controllers file. 
Also
```
Note that on many modern systems, [systemd(1)](https://man7.org/linux/man-pages/man1/systemd.1.html) automatically mounts
       the cgroup2 filesystem at /sys/fs/cgroup/unified during the boot
       process.
```
https://man7.org/linux/man-pages/man7/cgroups.7.html
We favour v1 because the subsystem files are in the v1 hierarchy in these cases, e.g. `/sys/fs/cgroup/memory/user.slice/user-0.slice/session-20778.scope/` (cgroup name obtained from /proc/self/cgroup) contains the `memory.limit_in_bytes` file among others while `/sys/fs/cgroup/unified//user.slice/user-0.slice/session-20778.scope/` doesn't